codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttEncoder.java

```diff
@@ -196,21 +223,34 @@ private static int getConnVariableHeaderFlag(MqttConnectVariableHeader variableH
     }
 
     private static ByteBuf encodeConnAckMessage(
+            ChannelHandlerContext ctx,
             ByteBufAllocator byteBufAllocator,
             MqttConnAckMessage message) {
-        ByteBuf buf = byteBufAllocator.buffer(4);
+        final MqttVersion mqttVersion = MqttCodecUtil.getMqttVersion(ctx);
+        ByteBuf propertiesBuf = encodePropertiesIfNeeded(mqttVersion,
+                byteBufAllocator,
+                message.variableHeader().properties());
+
+        ByteBuf buf = byteBufAllocator.buffer(4 + propertiesBuf.readableBytes());
         buf.writeByte(getFixedHeaderByte1(message.fixedHeader()));
-        buf.writeByte(2);
+        writeVariableLengthInt(buf, 2 + propertiesBuf.readableBytes());
         buf.writeByte(message.variableHeader().isSessionPresent() ? 0x01 : 0x00);
         buf.writeByte(message.variableHeader().connectReturnCode().byteValue());
+        buf.writeBytes(propertiesBuf);
 
         return buf;
     }
 
     private static ByteBuf encodeSubscribeMessage(
+            ChannelHandlerContext ctx,
             ByteBufAllocator byteBufAllocator,
             MqttSubscribeMessage message) {
-        int variableHeaderBufferSize = 2;
+        MqttVersion mqttVersion = MqttCodecUtil.getMqttVersion(ctx);
+        ByteBuf propertiesBuf = encodePropertiesIfNeeded(mqttVersion,
```

Comment:
you will need to call `propertiesBuf.release()` later as you not transfer ownership